### ISSUE TITLE
PDF internal links

### ISSUE THREAD

**Issue by robertosanval** [astefanutti/decktape#49]
I've a reveal presentation that exports very well with this command:

`$ ./bin/phantomjs decktape.js reveal http://reveal.local/ reveal.pdf`

The problem is on the internal PDF links that opens the browser to go to the original content.

What I'd like is that the internal links of the presentation converts to internal links in PDF.

Is that possible? I'm doing something wrong?

My HTML links are like this one:

`<a href="#/section-01">Internal example link</a>`

Thanks

**Comment by astefanutti** [comment]
Hyperlinks are currently rendered as PDF link annotation with the _URI action_ irrespective whether they happen to be _internal_ links or not. This has been implemented as part of DeckTape development with Vitallium/qtwebkit@ef91a25 and Vitallium/qtbase@d50c481. This should be possible to detect _internal_ links and have them rendered as `GoTo` action instead of `URI`, as defined in the PDF specification. I think [wkhtmltopdf](http://wkhtmltopdf.org) already provides something similar so that could be a start to provide support for it.

**Comment by robertosanval** [comment]
Thanks @astefanutti what could be the cost to implement that enhancement? I'm very interested on it.

**Comment by jonathanlilly** [comment]
I agree, this would be a very useful feature!

**Comment by gtolomei** [comment]
Hello,

Sorry for jumping onboard this discussion, which apparently seems already closed.
Thanks to the strong support of @astefanutti  I finally managed to generate good-looking pdf from RISE slideshow in Jupyter Notebook.
Here's the command I'm using at the moment to achieve that task:
`> decktape rise http://localhost:8888/notebooks/{NOTEBOOK_NAME}.ipynb?token={TOKEN} --load-pause 5000 -s 1920x1080 ~/path/to/outputfile.pdf`
Everything seems to work fine except for the hyperlinks included in my original slides. Those are correctly rendered in the final pdf (i.e., they appear as blue-colored and underlined) but unfortunately **they are not clickable**. 
I've tried opening the generated pdf both with Preview and Acrobat Reader but in none of them hyperlinks work fine. That's why I'm guessing there might be something odd while processing those hyperlinks.
Instead, hyperlinks work smoothly when presenting my RISE slideshow on the web browser.
Any help here would be much appreciated.

Thanks Kindly,
Gabriele

**Comment by astefanutti** [comment]
@gtolomei I've added support for hyperlinks in version `2.2.0`. Maybe you'll need to upgrade. This works for hyperlinks but not _internal_ links.

**Comment by gtolomei** [comment]
Thanks a lot @astefanutti for your quick reply!
I've just updated `decktape` to the latest version (2.2.2) and (external) hyperlinks work like a charm!

P.S.: In my RISE slideshow I have some Markdown cells which are left "unrun" on purpose (i.e., suppose you want to show the actual content of a Markdown cell on a slide and in the next slide you show how this actually renders); however, it seems decktape needs to run all the cells to produce the final pdf. Essentially, this leads to duplicate slides. Is there any overcome to this you may be aware of? Otherwise, I guess I will just skip the "unrun" cells.

Thanks a lot again,
Gabriele

**Comment by astefanutti** [comment]
The RISE plugin just iterate over the produced presentation by RISE. Are those _unrun_ cells supported by RISE?

**Comment by gtolomei** [comment]
Yes, they are. I mean, when I play my RISE slideshow I'm able to see sildes containing the "unrun" cells.

**Comment by astefanutti** [comment]
Hum I would have imagine DeckTape just exports them as it then. Isn't that the case?

**Comment by gtolomei** [comment]
Just to give you an example, below are two adjacent RISE slides:
<img width="1428" alt="screen shot 2017-09-13 at 4 40 55 pm" src="https://user-images.githubusercontent.com/9001735/30383666-7e1ec7ae-98a2-11e7-81ed-e80be3b4f294.png">
<img width="1423" alt="screen shot 2017-09-13 at 4 41 10 pm" src="https://user-images.githubusercontent.com/9001735/30383680-84ccd6e0-98a2-11e7-91b6-c5e0eedb4331.png">

**Comment by gtolomei** [comment]
Unfortunately, the corresponding PDF created with DeckTape just duplicates the "run" cells (i.e., in the example above, I have two pdf pages containing the second, rendered screenshot).

**Comment by astefanutti** [comment]
Interesting. Maybe we can move that discussion to a new issue then if you don't mind.

**Comment by gtolomei** [comment]
Yes, sure! In fact, I'm sorry I've used this thread but I thought it was something quick to fix!

**Comment by gtolomei** [comment]
Do you want me to fork this thread or have you already moved this to a different, new issue?
G.

**Comment by astefanutti** [comment]
Feel free to create a new issue. I'll clean that one then if that works for you.

**Comment by gtolomei** [comment]
Done! Here's the link to the [new issue](https://github.com/astefanutti/decktape/issues/111).
Thanks,
Gabriele

**Comment by astefanutti** [comment]
Issue with the RISE plugin followed-up in #111.

**Comment by phwebi** [comment]
> Hyperlinks are currently rendered as PDF link annotation with the _URI action_ irrespective whether they happen to be _internal_ links or not. This has been implemented as part of DeckTape development with [vitallium/qtwebkit@ef91a25](https://github.com/vitallium/qtwebkit/commit/ef91a25) and [vitallium/qtbase@d50c481](https://github.com/vitallium/qtbase/commit/d50c481). This should be possible to detect _internal_ links and have them rendered as `GoTo` action instead of `URI`, as defined in the PDF specification. I think [wkhtmltopdf](http://wkhtmltopdf.org) already provides something similar so that could be a start to provide support for it.

@astefanutti  I'm interested in taking a stab at implementing this and have read up on the PDF 2.0 specs for the GOTO action and named destinations. However, it's been a while since this issue has been opened and I see that you've since migrated to pupeteer. Any guidance on which dependencies I'd need to look at in order to implement this?
